Initial function call support in static analysis

spernsteiner opened the pull request:
This branch adds basic interprocedural analysis for pointer permissions.  It propagates the "easy"/dataflow permissions, such as `WRITE` and `OFFSET_ADD`, across function calls and returns.  For example, given `fn f(x: *mut i32) { *x = 1; }`, it will correctly infer that a call `f(p)` means that `p` must have `WRITE` permission, even if there is no explicit write in the calling function.

Handling of the `UNIQUE` permission is much more complex and is left for a separate branch.  In short, at each function call, the borrow checker needs to know how the lifetimes of the various pointers in the arguments and return values are related so that it knows how the call affects borrows within the caller.  For example, given `fn g(x: *mut i32) -> *mut i32 { x }`, it needs to know that the returned pointer borrows from the argument so it can properly remove `UNIQUE` for code like `let y = g(x); let z = g(x); *y = 1;`

Support for struct fields is also left for later, though that should be straightforward - pointers in field types simply get global `PointerId`s like function arguments do and participate in the dataflow analysis the same way.

kkysen commented on c2rust-analyze/src/context.rs:
```diff
@@ -55,10 +58,18 @@ pub use crate::pointer_id::PointerId;
 pub type LTy<'tcx> = LabeledTy<'tcx, PointerId>;
 pub type LTyCtxt<'tcx> = LabeledTyCtxt<'tcx, PointerId>;
 
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub struct LFnSig<'tcx> {
```
Just `L` for `Labeled` seems a bit short here.

spernsteiner replied:
`LTy` is used pervasively throughout this codebase, so I don't think it needs a long name.  And `LFnSig` is named by analogy with rustc's ordinary `FnSig` and `Ty`.

kkysen replied:
Ok, I think it's fine as long as `L` only means `Labeled`, but if it means `locals` or anything else elsewhere, then it gets very confusing.

kkysen commented on c2rust-analyze/src/dataflow/type_check.rs:
```diff
@@ -209,16 +213,50 @@ impl<'tcx> TypeChecker<'tcx, '_> {
+            None => todo!("call to unknown function {:?}", def_id),
+        };
+        if substs.non_erasable_generics().next().is_some() {
+            todo!("call to generic function {:?} {:?}", def_id, substs);
```
```suggestion
            todo!("call to generic function {def_id:?} {substs:?}");
```

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -35,29 +41,113 @@ mod pointer_id;
+
+impl<T> MaybeUnset<T> {
+    pub fn set(&mut self, x: T) {
+        if self.0.is_some() {
+            panic!("value is already set");
+        }
+        self.0 = Some(x);
```
```suggestion
        if self.0.replace(x).is_some() {
            panic!("value is already set");
        }
```

spernsteiner replied:
IMO this makes the logic harder to follow

kkysen replied:
I disagree.  The `Option` methods are designed for exactly this and are idiomatic.

aneksteind replied:
fwiw I agree with Stuart that this is harder to follow -- for me it's because I'm less familiar with `replace`

kkysen replied:
There's also `std::mem::{replace,take}`.  It's an idiom Rust uses in multiple places, so I do think it's still worth it to use.  You could use `match` instead of `map` or `expect`, too, as the logic is easier for a beginner to follow, but you would probably encourage people to use the dedicated method and have them learn it.

spernsteiner replied:
The connotation of `replace` to me is that there are two interesting values, a new one and (possibly) an old one.  Here, we don't care about the old value at all (and in fact we're asserting that there isn't one), so `replace` seems like an unusual choice.

Similarly, `take` is about extracting the value out of an option, so using it when we don't care about the value is a bit weird.  I would guess that this `x.take().unwrap();` type of construct (where the result is unused) is almost unheard of in Rust code.

Overall, I think the original versions of these functions express the high-level intent about as clearly and directly as they possibly could.  Switching to `replace`/`take` brings in concepts of old/new values which are not actually helpful for understanding.

kkysen replied:
[`Cell::set` is implemented](https://doc.rust-lang.org/src/core/cell.rs.html#362) using `Cell::replace`.  That's exactly what we're doing here, using replace to set a new value.  Except unlike `Cell`, we don't care about propagating the old value to the user, but we want to make sure we didn't actually replace a value, so we panic instead.

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -35,29 +41,113 @@ mod pointer_id;
+    }
+
+    pub fn clear(&mut self) {
+        if self.0.is_none() {
+            panic!("value is already cleared");
+        }
+        self.0 = None;
```
```suggestion
        if self.0.take().is_none() {
            panic!("value is already cleared");
        }
```

spernsteiner replied:
IMO this makes the logic harder to follow

kkysen replied:
I think we also get a better, more standard, more recognizable and related to `Option` error message here with `expect`.

spernsteiner replied:
I hadn't considered the panic message.  IMO, the fact that the message would mention `Option` is a (very minor) downside of the `expect` version.  The user of `MaybeUnset<T>` should not be thinking of it as an `Option<T>`.  They should think of it as a `T`, with the minor caveat that Rust's normal static checks for variable initialization checks don't apply.

kkysen replied:
Here I see `take` as in it's taking away the current value, the same as clearing it.

spernsteiner replied:
This simply is not an idiomatic use of `take`.  The primary purpose of `take` is to get the old value.  This is reflected in the docs ("Takes the value out of the option") and in common usage.  I did a quick grep of rust-lang/rust, and every instance of `take()`, `take().unwrap()`, or `take().expect()` that I found (outside of test code) was actually using the extracted value for something afterwards.

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -35,29 +41,113 @@ mod pointer_id;
+    struct FuncInfo<'tcx> {
+        acx_data: MaybeUnset<AnalysisCtxtData<'tcx>>,
+        dataflow: MaybeUnset<DataflowConstraints>,
+        l_equiv: MaybeUnset<LocalEquivSet>,
```
Previously `l` meant `Labeled`.  Here I assume it means `Local`.  I wouldn't use `l`/`L` for both of them.

spernsteiner replied:
OK, I renamed all the `l_`/`g_`s to `local_`/`global_`

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -35,29 +41,113 @@ mod pointer_id;
+            let lty = match mir.local_kind(local) {
+                LocalKind::Var | LocalKind::Temp => acx.assign_pointer_ids(decl.ty),
+                LocalKind::Arg => {
+                    debug_assert!(local.as_usize() >= 1 && local.as_usize() <= mir.arg_count);
```
Why a `debug_assert!` and not just an `assert!`?  If the assert fails but it's in release mode, this will cause an overflow, right?

spernsteiner replied:
If this condition doesn't hold, the next line will panic with index out of bounds (possibly after underflowing to `usize::MAX`).  The assertion here is technically redundant - its only purpose is to express the expected property in a more straightforward way.

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -73,40 +163,85 @@ fn run(tcx: TyCtxt) {
-        // Remap pointers based on equivalence classes, so all members of an equivalence class now
-        // use the same `PointerId`.
-        let (l_counter, l_equiv_map) = l_equiv.renumber(&g_equiv_map);
+    for ldid in tcx.hir().body_owners() {
+        let info = func_info.get_mut(&ldid).unwrap();
+        let (l_counter, l_equiv_map) = info.l_equiv.renumber(&g_equiv_map);
```
Why is `l` sometimes used without a separator (like `l*`) and sometimes with a `_` (like `l_*`)?

spernsteiner replied:
`ldid` is (roughly) an initialism for `LocalDefId`, similar to `did` for `DefId`.  The latter appears pervasively within rustc.

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -73,40 +163,85 @@ fn run(tcx: TyCtxt) {
+    for ldid in tcx.hir().body_owners() {
+        let info = func_info.get_mut(&ldid).unwrap();
+        let (l_counter, l_equiv_map) = info.l_equiv.renumber(&g_equiv_map);
         eprintln!("l_equiv_map = {:?}", l_equiv_map);
```
```suggestion
        eprintln!("l_equiv_map = {l_equiv_map:?}");
```

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -73,40 +163,85 @@ fn run(tcx: TyCtxt) {
+    let order = walk_callgraph(tcx);
+    eprintln!("callgraph traversal order:");
+    for &ldid in &order {
+        eprintln!("  {:?}", ldid);
```
```suggestion
        eprintln!("  {ldid:?}");
```

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -180,11 +315,36 @@ fn run(tcx: TyCtxt) {
-        _ => PointerId::NONE,
-    })
+trait AssignPointerIds<'tcx> {
+    fn lcx(&self) -> LTyCtxt<'tcx>;
+    fn new_pointer(&mut self) -> PointerId;
```
```suggestion
    fn lcx(&self) -> LTyCtxt<'tcx>;
    
    fn new_pointer(&mut self) -> PointerId;
```

kkysen replied:
Why was this resolved?  There's spacing between the other methods in this trait.

kkysen replied:
Oh wait, I think I was looking at an outdated version.

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -222,6 +382,68 @@ fn describe_span(tcx: TyCtxt, span: Span) -> String {
+        f: F,
+    }
+
+    impl<'tcx, F: FnMut(LocalDefId)> Visitor<'tcx> for CalleeVisitor<'_, 'tcx, F> {
+        fn visit_operand(&mut self, operand: &Operand<'tcx>, _location: Location) {
+            let ty = operand.ty(self.mir, self.tcx);
+            if let Some(Callee::Other { def_id, .. }) = util::ty_callee(self.tcx, ty) {
+                if let Some(ldid) = def_id.as_local() {
+                    (self.f)(ldid);
+                }
+            }
+        }
+    }
```
I didn't realize you can `impl` inside of a function.  Neat!

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -222,6 +382,68 @@ fn describe_span(tcx: TyCtxt, span: Span) -> String {
+            continue;
+        }
+        stack.push(Visit::Pre(root_ldid));
+        while let Some(x) = stack.pop() {
+            match x {
```
```suggestion
        while let Some(ldid) = stack.pop() {
            match ldid {
```

spernsteiner replied:
`x` is a wrapper and not (just) a `LocalDefId`, so naming it `ldid` would be misleading IMO

kkysen replied:
I think `ldid` makes more sense than `x`.  It always only contains an `ldid`, it's just either `Pre` or `Post`.

spernsteiner replied:
This function takes a recursive algorithm (postorder graph traversal) and implements it iteratively.  `Visit` is best thought of not as "a `LocalDefId` plus an extra bit" (even though concretely that's exactly what it is), but as "a step of the original recursive algorithm".

kkysen replied:
Can it be named `visit` then?

spernsteiner replied:
OK, renamed to `visit`

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -222,6 +382,68 @@ fn describe_span(tcx: TyCtxt, span: Span) -> String {
 
+/// Return all `body_owners`, ordered according to a postorder traversal of the graph of references
+/// between bodies.
+fn walk_callgraph(tcx: TyCtxt) -> Vec<LocalDefId> {
```
Could be an iterator, external like `Iterator` or internal like `for_each_callee`.

spernsteiner replied:
In the future (in code I've already written based on top of this branch), it seems like we'll need to iterate over the ordering multiple times, so I think leaving it as `Vec` is best.

kkysen replied:
I think `walk` isn't the best name then.  How about `collect` or `walk_and_collect`?  `walk` sounds like an iterator (internal or external).

spernsteiner replied:
OK, I've renamed it to `body_owners_postorder` (since it returns the same sequence as `tcx.body_owners()`, but sorted according to a postorder traversal)

kkysen commented on c2rust-analyze/tests/filecheck/call1.rs:
```diff
@@ -0,0 +1,48 @@
+// With a naive iteration order, reaching a fixpoint used to take 3 iterations.  By following a
+// postorder traversal of the callgraph, we reduce that to 2 iterations: the first iteration
+// computes all the right permissions, and the second checks that we've actually reached a
+// fixpoint.
```
Does it make sense for these to be doc comments?

spernsteiner replied:
This comment is describing reasoning for the `// CHECK` below, not a function or module.  Plus, I can't imagine anyone ever running `rustdoc` on these files.

kkysen replied:
Okay, that's reasonable.

aneksteind commented on c2rust-analyze/src/context.rs:
```diff
@@ -247,6 +261,23 @@ impl<'tcx> AnalysisCtxtData<'tcx> {
+    }
+}
+
+fn remap_lty_pointers<'tcx, T>(lcx: LTyCtxt<'tcx>, map: &T, lty: LTy<'tcx>) -> LTy<'tcx>
```
would you mind adding a docstring here explaining the use case

aneksteind commented on c2rust-analyze/src/dataflow/type_check.rs:
```diff
@@ -48,16 +49,25 @@ impl<'tcx> TypeChecker<'tcx, '_> {
     }
 
     pub fn visit_place_ref(&mut self, pl: PlaceRef<'tcx>, ctx: PlaceContext) -> LTy<'tcx> {
+        let mutbl = match ctx {
```
do you think this would be useful as its own utility function?

spernsteiner replied:
It turns out the use of `PlaceContext` was a holdover from previous code that was implemented as a `mir::Visitor`.  I changed these functions to take `mutbl: Mutability` directly instead.

aneksteind commented on c2rust-analyze/src/dataflow/type_check.rs:
```diff
@@ -48,16 +49,25 @@ impl<'tcx> TypeChecker<'tcx, '_> {
                 ProjectionElem::Deref => {
                     // All derefs except the last are loads, to retrieve the pointer for the next
-                    // deref.  The last deref is either a load or a store, depending on `ctx`.
+                    // deref.  However, if the last deref is `&mut` (and is used mutably), then the
```
is the last deref being &mut and used mutably determined exclusively by `mutbl`?

spernsteiner replied:
Yes - I updated the comment slightly

aneksteind reviewed (approved):
a general comment -- i think more documentation and longer variable names would have helped my understanding without having to scroll up and down for more context. any additions along those lines would be welcome

spernsteiner commented:
I think I've addressed all the previous comments.  I used the "resolve conversation" button liberally to help keep track of which threads I'd already handled - feel free to unresolve if there's more to discuss.

kkysen commented on c2rust-analyze/src/dataflow/type_check.rs:
```diff
@@ -117,14 +112,8 @@ impl<'tcx> TypeChecker<'tcx, '_> {
-                let ctx = PlaceContext::NonMutatingUse(NonMutatingUseContext::Move);
-                self.visit_place(pl, ctx)
-            }
+            Operand::Copy(pl) => self.visit_place(pl, Mutability::Not),
+            Operand::Move(pl) => self.visit_place(pl, Mutability::Not),
```
Could these match arms be combined?

kkysen commented on c2rust-analyze/src/main.rs:
```diff
@@ -66,47 +169,107 @@ fn run(tcx: TyCtxt) {
+    // Remap pointers based on equivalence classes, so all members of an equivalence class now use
+    // the same `PointerId`.
+    let (global_counter, global_equiv_map) = global_equiv.renumber();
+    eprintln!("global_equiv_map = {:?}", global_equiv_map);
```
```suggestion
    eprintln!("global_equiv_map = {global_equiv_map:?}", );
```

kkysen reviewed (changes requested):
Can you make the referenced items in doc comments into real intra-doc links?  It helps a lot with navigating documentation and ensuring they stay up-to-datr through refactors.

Also, I think you should use the existing, idiomatic `Option` methods when the logic is exactly what those methods do.  If not using them was not as clear, then why are they in `core`?

There's also a few other small suggestions I made.

spernsteiner commented:
I've addressed the remaining comments.  I added doc links only in the comments that were added or modified in this PR, to avoid polluting the diff with unrelated changes - the rest can be done in a separate PR if needed.
